### DISCUSSION TITLE
Get `decorated` page endpoint

### DISCUSSION THREAD

**Discussion by WSH032** [zauberzeug/nicegui#2382] [Ideas / Feature Requests]
Hi, thank you for developing this fantastic library. 💖

---

## Question

During usage, I notice that `nicegui` wraps the original endpoint function, making it impossible to retrieve the actual endpoint from the `scope` of request

```py
from fastapi import Request
from nicegui import ui


@ui.page("/")
def foo(request: Request):
    endpoint = request.scope.get("endpoint")
    print(endpoint)
    print(endpoint is foo)  # False

if __name__ == "__main__":
    ui.run(reload=False)
```

---

## Feature Requests

How about adding a `page_endpoints` property for `class nicegui.page.page`,

https://github.com/zauberzeug/nicegui/blob/ccab1812fd414080f0f844827b2f7d8ea1daf77f/nicegui/page.py#L134

something like:

```py
_OriginEndpointType = Callable[..., Any]
_DecoratedEndpointType = Callable[..., Coroutine[Any, Any, Response]]

def __init__(self, ...):
    ...
    self.page_endpoints: Dict[_OriginEndpointType, _DecoratedEndpointType] = {}

def __call__(self, ...):
    ...
    return self.page_endpoints[func] = decorated
```

**Comment by rodja** [comment]
I wonder for what purpose you need the actual undecorated endpoint?

**Reply by WSH032** [reply]
I have a dependency for authentication, and I need to make a decision based on the endpoint the dependency is handling regarding whether to intercept unauthorized requests.

https://github.com/WSH032/aria2-server-gui/blob/c134d7ff8b0b5744e3e9acaa805ab266cf0a1c16/src/aria2_server/app/core/_auth/_dependencies.py#L109-L125

---

This is a minimal example:
```py
from nicegui import ui
from fastapi.requests import HTTPConnection
from fastapi import HTTPException, Depends
from typing import Optional


class User:
    def __init__(self):
        self._private_page = set()

    def private_page(self, endpoint):
        self._private_page.add(endpoint)
        return endpoint

    def __call__(self, conn: HTTPConnection) -> Optional[str]:
        user = conn.cookies.get("user")
        if conn.scope.get("endpoint") in self._private_page:
            if user is None:
                raise HTTPException(307, headers={"Location": "/login"})
        return user


user = User()


@ui.page("/")
def index(user: str = Depends(user)):
    ui.label(f"Hello {user}!")


@ui.page("/login")
def login(user: str = Depends(user)):
    ui.label(f"Hello {user}!")

# Here, `ui.page` indeed returns a endpoint,
# but the endpoint isn't the actual the page endpoint.
@user.private_page
@ui.page("/private_page")
def private_page(user: str = Depends(user)):
    ui.label(f"Hello {user}!")

```

**Reply by rodja** [reply]
Hmm... what is the expected outcome? And what do you see instead? When I run the code all three pages just show the label "Hello None!".

**Reply by WSH032** [reply]
> When I run the code all three pages just show the label "Hello None!".

Because `@user.private_page` did not get the actual page endpoint.

What I want to implement is accessing `/private_page` will be redirected to `/login`.

**Edit:**

`if conn.scope.get("endpoint") in self._private_page:`

Here, we can never obtain the actual page endpoint, so the redirection will not occur.

**Reply by WSH032** [reply]
Calling `conn.scope.get("endpoint")`, we can only get `async def decorated(*dec_args, **dec_kwargs) -> Response:` 
https://github.com/zauberzeug/nicegui/blob/ccab1812fd414080f0f844827b2f7d8ea1daf77f/nicegui/page.py#L95

---

However, we cannot get this actual endpoint (`async def decorated(*dec_args, **dec_kwargs) -> Response:`) in advance, because `ui.page` neither returns it nor retains a reference to it.

Therefore, we are unable to add this actual endpoint to `user._private_page`.

**Reply by rodja** [reply]
Ah, now I understand! Thanks for elaborating @WSH032. I think we could solve it by adding `@wraps(func)` to the `async def decorated` function in `page.__call__`. With this change the original functions metadata a preserved and you can write:

```py
    def __call__(self, conn: HTTPConnection) -> Optional[str]:
        user = conn.cookies.get("user")
        endpoint = conn.scope.get("endpoint")
        if endpoint is not None and endpoint.__wrapped__ in self._private_page:
            if user is None:
                print('redirecting')
                raise HTTPException(307, headers={"Location": "/login"})
        return user
```

What do you think? It would be great if you could create a pull request and thereby become official contributor.

**Reply by WSH032** [reply]
I do agree with using `@wraps(func)`; it can greatly improve Swagger docs.

```py
@ui.page("/")
def index():
    """The index page."""
    ui.label("Hello")
```

For example, in the case of this code, previously, it could only display `decorated` endpoints without documentation.

![P(QO(5V(XZXPYOC(WAYZ`QQ](https://github.com/zauberzeug/nicegui/assets/126865849/058bce39-a55d-4e3c-ab93-85a02138adf6)


After using `@wraps(func)`, it is now able to display the actual `index` endpoint documentation.

![MPR~5 ZOQ{Q(LHL8@F_J@O](https://github.com/zauberzeug/nicegui/assets/126865849/2fe3fbe8-ad31-4730-b0df-e9f4fbd4c564)

---

However, for my use case, I still prefer an API similar to `self.page_endpoints: Dict[_OriginEndpointType, _DecoratedEndpointType] = {}`.

Authentication is not only required for nicegui's page endpoint but also for pure FastAPI endpoints, for example:

```py
@app.get("/")
def index(user: str = Depends(user)):
    return "secret"
```
So, we need to execute code like this:
```py
if endpoint in endpoint_set or (
        hasattr(endpoint, "__wrapped__") and endpoint.__wrapped__ in endpoint_set
    ):
```

<details>
  <summary>This introduces a performance cost</summary>

  ```python
import functools


def endpoint():
    return 0


@functools.wraps(endpoint)
def decorated_endpoint():
    return endpoint()


endpoint_set = {(lambda i: i) for i in range(10)}


def need_to_check_wrapped(endpoint):
    if endpoint in endpoint_set or (
        hasattr(endpoint, "__wrapped__") and endpoint.__wrapped__ in endpoint_set
    ):
        return True


def no_need_to_check_wrapped(endpoint):
    if endpoint in endpoint_set:
        return True


%timeit need_to_check_wrapped(decorated_endpoint) ; need_to_check_wrapped(endpoint)
%timeit no_need_to_check_wrapped(decorated_endpoint) ; no_need_to_check_wrapped(endpoint)
```
```console
322 ns ± 5.34 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
201 ns ± 1.97 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
</details>

Another consideration is:
- what if other users still want to obtain decorated endpoints in advance for other purposes?
- In the authentication example provided earlier, the code logic was written by me, so I could easily modify the code. But what if it's another third-party authentication library?

**Reply by WSH032** [reply]
Using [weakref.WeakKeyDictionary](https://docs.python.org/3/library/weakref.html?highlight=weakref#weakref.WeakKeyDictionary), we can introduce this new API with minimal performance loss. Additionally, deprecating it is straightforward. All that needs to be done is:
```diff
- self.page_endpoints[func] = decorated
+ self.page_endpoints[func] = func
```

---

You have been maintaining NiceGUI for a long time, and I may not have the same foresight in API design and maintenance as you do. I apologize if I am making too many requests.

The solution you provided already meets my needs. I am happy to initiate a pull request. Please give me some time to study the `CONTRIBUTING.md` guidelines.

Let me know your thoughts, thank you!

**Reply by rodja** [reply]
I just created pull request #2412 to add the `@wraps` decorator.

@WSH032 the diff is surely bigger than changing `decorated` to `func` because we do not have `page_endpoints` yet. While your arguments are sound, I'm still not sure if it's the best approach. Could you maybe create a pull request including an example of the usage? That would help a lot.

**Reply by WSH032** [reply]
After merging #2412, I came up with a compromise:

```py
from collections import defaultdict

from nicegui import APIRouter, ui
from fastapi.routing import APIRoute

private_router = APIRouter()


@private_router.page("/private")
def private_page():
    ui.label("secret")


def build_endpoint_map(router: APIRouter):
    map_ = defaultdict(list)

    for route in router.routes:
        if isinstance(route, APIRoute) and hasattr(route.endpoint, "__wrapped__"):
            map_[route.endpoint.__wrapped__].append(route.endpoint)

    return map_

page_endpoints = build_endpoint_map(private_router)
```

This way, we maintain our own `page_endpoints`.

---

This approach still has some drawbacks:

- Lack of type annotation support for `__wrapped__`.
    ![BE J2VUYWM@ITF 9(YS4E1](https://github.com/zauberzeug/nicegui/assets/126865849/a871e269-0f11-4259-acf5-fd28ee9d81d7)

- `@wraps` is an internal implementation detail of NiceGUI, and **it seems it should not be considered a stable API?**

---
> Could you maybe create a pull request including an example of the usage?

I haven't come up with a better solution either. For me, the current solution completely satisfies my use case, so it seems we should close this discussion for now?

Thank you very much for your assistance!
